src/lib.rs

```diff
@@ -174,8 +176,70 @@ fn enum_variant_type_impl(ast: DeriveInput) -> proc_macro2::TokenStream {
     let data_enum = data_enum(&ast);
     let variants = &data_enum.variants;
 
+    let mut wrap_in_module = None::<Ident>;
+    let mut derive_for_all_variants = None::<Attribute>;
+    let mut marker_trait_ident = None::<Ident>;
+
+    for attr in ast.attrs.iter() {
+        if attr.path.is_ident("evt") {
+            if let Ok(Meta::List(list)) = attr.parse_meta() {
+                for item in list.nested.iter() {
+                    match item {
+                        NestedMeta::Meta(Meta::NameValue(name_value)) => {
+                            if let (true, Lit::Str(lit_str)) =
+                                (name_value.path.is_ident("module"), &name_value.lit)
+                            {
+                                wrap_in_module =
+                                    Some(Ident::new(&lit_str.value(), Span::call_site()));
+                            } else {
+                                panic!("Expected #[evt(module = \"some_module_name\")]");
+                            }
+                        }
+                        NestedMeta::Meta(Meta::List(list)) => {
+                            if list.path.is_ident("derive") {
+                                let items = list
+                                    .nested
+                                    .iter()
+                                    .map(|nested_meta| {
+                                        if let NestedMeta::Meta(Meta::Path(path)) = nested_meta {
+                                            path.clone()
+                                        } else {
+                                            panic!("Expected #[evt(derive(Clone, Debug))]");
+                                        }
+                                    });
+                                derive_for_all_variants = Some(parse_quote! {
+                                    #[derive( #(#items),* )]
+                                });
+                            } else if list.path.is_ident("derive_marker_trait") {
+                                if list.nested.len() != 1 {
+                                    panic!("Expected #[evt(derive_marker_trait(SingleMarkerTrait))]");   
+                                }
+                                let ident = if let NestedMeta::Meta(Meta::Path(path)) = 
+                                    list.nested.iter().next().unwrap() {
+                                    path.get_ident().unwrap().clone()
+                                } else {
+                                    panic!("Expected #[evt(derive_marker_trait(SingleMarkerTrait))]");   
+                                };
+                                marker_trait_ident = Some(ident);
+                            }
+                        }
+                        _ => panic!("Invalid evt attr"),
+                    }
+                }
+            } else {
+                panic!("Invalid evt attr")
```

Comment:
hm, it's been a while since I've done proc macros, but does rust filter out attributes that aren't listed in the `proc_macro_derive`?

if so, then we can do something like this:

```rust
panic!("`EnumVariantType` does not recognize `{}` as an attribute. Did you mean `evt`?")
```

and if not, we should probably just ignore unrecognized attributes -- I remember something about "proc macros can see all other attributes except `derive`s"

Reply:
Interesting point, just checked and no rust does not filter attributes not specified in the attribute argument on the `proc_macro_derive`. For:

```rust
#[derive(Debug, EnumVariantType)]
#[evt(...)]
#[rustfmt::ignore]
struct A;
```
The following attributes are present

```
ast.attrs = [
    Attribute {
        path: Path {
            leading_colon: None,
            segments: [
                PathSegment {
                    ident: Ident {
                        ident: "evt",
        ...
    },
    Attribute {
        path: Path {
            leading_colon: None,
            segments: [
                PathSegment {
                    ident: Ident {
                        ident: "rustfmt",
        ...
    },
]
```

And I think this should already be handled as there isn't anything happening on not "evt" attributes. The panic is only if the evt is not a meta list e.g. `#[evt]` or `#[evt = "invalid_use"]`
```rust
for attr in ast.attrs.iter() {
    if attr.path.is_ident("evt") {
        if let Ok(Meta::List(list)) = attr.parse_meta() {
            ...
        } else {
            panic!("Invalid evt attr")
        }
    }
}
```

I couldn't think of a better error message 😄 Maybe something which references correct usage or links to it. Also maybe should be returning a `compile_error!` macro rather than panicking. Will have a look at tidying my error logic up 👍

Reply:
> And I think this should already be handled as there isn't anything happening on not "evt" attributes. The panic is only if the evt is not a meta list e.g. #[evt] or #[evt = "invalid_use"]

oh yes, my eyes didn't parse that right (though this was at the 2-indent level) :bowing_man: